presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/Util.java

```diff
@@ -44,10 +45,25 @@ private Util()
      * <p>
      * If all inputs are used, return Optional.empty() to indicate that no pruning is necessary.
      */
-    public static Optional<Set<VariableReferenceExpression>> pruneInputs(Collection<VariableReferenceExpression> availableInputs, Collection<Expression> expressions, TypeProvider types)
+    public static Optional<Set<VariableReferenceExpression>> pruneInputs(
+            Collection<VariableReferenceExpression> availableInputs,
+            Collection<RowExpression> expressions,
+            TypeProvider types)
     {
         Set<VariableReferenceExpression> availableInputsSet = ImmutableSet.copyOf(availableInputs);
-        Set<VariableReferenceExpression> prunedInputs = Sets.filter(availableInputsSet, VariablesExtractor.extractUnique(expressions, types)::contains);
+        Set<VariableReferenceExpression> referencedInputs;
+        if (expressions.stream().anyMatch(OriginalExpressionUtils::isExpression)) {
+            // TODO remove once all pruneInputs rules are below translateExpressions.
+            // We should either contain all original expression or all translated row expression.
```

Comment:
I'd use `allMatch` in that case and remove the comment. That way the logic explains itself. (Or maybe add comment to `else` saying that otherwise we expect all real row expression.

Reply:
Thanks, I will check if all or none and throw an explict exception when there is mixed expression and row expression.